pkg/util/metamorphic/constants.go

```diff
@@ -161,23 +110,171 @@ func ConstantWithTestBoolWithoutLogging(name string, defaultValue bool) bool {
 	return constantWithTestBoolInternal(name, defaultValue, false /* doLog */)
 }
 
+func constantWithTestBoolInternal(name string, defaultValue bool, doLog bool) bool {
+	return getConstantInternal(name,
+		defaultValue,
+		strconv.ParseBool,
+		metamorphicBoolProbability,
+		func(*rand.Rand) bool { return !defaultValue },
+		doLog)
+}
+
 // ConstantWithTestChoice is like ConstantWithTestValue except it returns a
 // random choice (equally weighted) of the given values. The default value is
 // included in the random choice.
 //
 // The given name is used for logging.
 func ConstantWithTestChoice[T any](name string, defaultValue T, otherValues ...T) T {
+	return getConstantInternal(name,
+		defaultValue,
+		func(s string) (T, error) {
+			v, err := parseChoice[T](s)
+			return v.(T), err
+		},
+		1.0,
+		func(r *rand.Rand) T {
+			values := append([]T{defaultValue}, otherValues...)
+			return values[rng.r.Int63n(int64(len(values)))]
+		},
+		true)
+}
+
+// getConstantInternal returns a value of type T for the given name. If the name
+// has been specified as an override, the parseValue func is used to parse the
+// user-provided value. If no such override is specified, it either returns the
+// defaultValue or a value generated by generateValue()
+func getConstantInternal[T any](
+	name string,
+	defaultValue T,
+	parseValue func(string) (T, error),
+	probability float64,
+	generateValue func(r *rand.Rand) T,
+	doLog bool,
+) T {
+	if val, ok := valueFromOverride(name, parseValue); ok {
+		logMetamorphicValue(name, val, override)
+		return val
+	}
+
 	if metamorphicutil.IsMetamorphicBuild {
-		values := append([]T{defaultValue}, otherValues...)
 		rng.Lock()
 		defer rng.Unlock()
-		value := values[rng.r.Int63n(int64(len(values)))]
-		logMetamorphicValue(name, value)
-		return value
+		if rng.r.Float64() < probability {
```

Comment:
nit: One consideration to consider to minimize the impact of setting a env constant is to call `generateValue` first and then call `valueFromOverride` afterwards and replace the value if it is set. This will mean that if you have a known bad seed and it runs through setting all the values with that seed if you later set an env variable, everything else will be unchanged.

Reply:
Great call. I restructured this so that we should get the same number of uses of the random number generate even when you set the override.  I manually tested that at least in the simple test I added this seems to work.